src/Extensions/Diagnostics.HealthChecks/Internal/HealthCheckPublisher.cs

```diff
@@ -0,0 +1,123 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Fabric;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Diagnostics.HealthChecks;
+using Microsoft.Extensions.ObjectPool;
+using ServiceFabricHealth = System.Fabric.Health;
+
+namespace Microsoft.Omex.Extensions.Diagnostics.HealthChecks
+{
+	internal abstract class HealthCheckPublisher : IHealthCheckPublisher
+	{
+		protected readonly ObjectPool<StringBuilder> StringBuilderPool;
+
+		internal static string HealthReportSourceId => nameof(HealthCheckPublisher);
+
+		internal const string HealthReportSummaryProperty = "HealthReportSummary";
+
+		public HealthCheckPublisher(ObjectPoolProvider objectPoolProvider) => StringBuilderPool = objectPoolProvider.CreateStringBuilderPool();
+
+		public abstract Task PublishAsync(HealthReport report, CancellationToken cancellationToken);
+
+		protected abstract ServiceFabricHealth.HealthInformation FinalizeHealthReport(string healthReportSummaryProperty, ServiceFabricHealth.HealthState healthState);
+
+		protected void PublishAllEntries(HealthReport report, Action<ServiceFabricHealth.HealthInformation> publishFunc,
+			 CancellationToken cancellationToken)
+		{
+			try
+			{
+				// We trust the framework to ensure that the report is not null and doesn't contain null entries.
+				foreach (KeyValuePair<string, HealthReportEntry> entryPair in report.Entries)
+				{
+					cancellationToken.ThrowIfCancellationRequested();
+					publishFunc(BuildSfHealthInformation(entryPair.Key, entryPair.Value));
+				}
+
+				cancellationToken.ThrowIfCancellationRequested();
+				publishFunc(BuildSfHealthInformation(report));
+			}
+			catch (FabricObjectClosedException)
+			{
+				// Ignore, the service instance is closing.
+			}
+		}
+
+
+		protected ServiceFabricHealth.HealthInformation BuildSfHealthInformation(HealthReport report)
+		{
+			int entriesCount = report.Entries.Count;
+			int healthyEntries = 0;
+			int degradedEntries = 0;
+			int unhealthyEntries = 0;
+			foreach (KeyValuePair<string, HealthReportEntry> entryPair in report.Entries)
+			{
+				switch (entryPair.Value.Status)
+				{
+					case HealthStatus.Healthy:
+						healthyEntries++;
+						break;
+					case HealthStatus.Degraded:
+						degradedEntries++;
+						break;
+					case HealthStatus.Unhealthy:
+					default:
+						unhealthyEntries++;
+						break;
+				}
+			}
+
+			StringBuilder descriptionBuilder = StringBuilderPool.Get();
```

Comment:
When you are done with the StringBuilder you should return it to the pool.

Reply:
I still don't see this builder returned to the pool, why thread resolved?

Reply:
Sorry, I was looking on old version